### PR TITLE
Logging and bug fixes. 

### PR DESCRIPTION
**Pull request by DanielKrawisz**
Using logs to generate experimental data so as to figure out why bmd is misbehaving. 

Bug fixes: 
* Increase maxKnownInventory to MaxInvPerMsg. This is necessary until we get the object manager to behave more intelligently because there are around 30,000 objects floating around the bitmessage network right now. 
* There was a bug in peerHandler that caused bmd not to connect to new outbound peers when it needed to. 
* Fixed bug in Connection which caused it not to return the correct address when it was not connected.
* Fixed bug that sometimes caused the server not to get the message that a peer had disconnected.
* Fixed bug that caused the peer not to clean up after itself if it failed to connect initially. 
* Fixed bug that caused the connection not to close properly if it returned an error while reading a message. 

BMD is now more stable, but still not stable enough to be usable.

### CODE REVIEW

File: peer.go [hunk 1]
```diff
@@ -118,6 +131,7 @@ func (p *bmpeer) ProtocolVersion() uint32 {
 // current state.
 func (p *bmpeer) PushVersionMsg() {
 	if p.versionSent {
+		peerLog.Infof(p.peer.PrependAddr("For some reason we are trying to send a version message a second time."))
```
Comment (ishbir) [inline peer.go]: Shouldn't this be a warning or higher, instead of info?

File: objectmanager.go [hunk 2]
```diff
@@ -118,6 +119,9 @@ func (om *ObjectManager) handleObjectMsg(omsg *objectMsg) {
 		// thus disconnecting legitimate peers. We want to prevent against such
 		// an attack by checking that objects came from peers that we requested
 		// from.
+		peerLog.Infof(omsg.peer.peer.PrependAddr(
+			fmt.Sprint("Disconnecting because of unrequested object ", 
+				invVect.Hash.String()[:8], " received.")))
```
Comment (ishbir) [inline objectmanager.go]: I think this would be better off as a warning.

File: objectmanager.go [hunk 3]
```diff
@@ -131,6 +135,7 @@ func (om *ObjectManager) handleObjectMsg(omsg *objectMsg) {
 		om.server.db.InsertObject(obj)
 	}
 
+	peerLog.Infof(omsg.peer.peer.PrependAddr(fmt.Sprint("Object ",invVect.Hash.String()[:8]," received.")))
```
Comment (ishbir) [inline objectmanager.go]: Should be a trace.

File: objectmanager.go [hunk 4]
```diff
@@ -174,6 +179,7 @@ func (om *ObjectManager) handleInvMsg(imsg *invMsg) {
 		return
 	}
 
+	peerLog.Infof(imsg.peer.peer.PrependAddr(fmt.Sprint(i," unknown hashes received.")))
```
Comment (ishbir) [inline objectmanager.go]: Should be a trace.

File: peer.go [hunk 5]
```diff
@@ -87,24 +87,37 @@ func (p *bmpeer) HandshakeComplete() bool {
 
 // disconnect disconnects the peer.
 func (p *bmpeer) disconnect() {
-	p.peer.Disconnect()
-
-	// Only tell object manager we are gone if we ever told it we existed.
-	if p.HandshakeComplete() {
-		p.server.objectManager.DonePeer(p)
+	peerLog.Debug(p.peer.PrependAddr("disconnect called."))
```
Comment (ishbir) [inline peer.go]: Debug or info? I'd personally go with info, based on the pattern followed in btcd. (e.g. server shutdown, database shutdown is info)

File: peer.go [hunk 6]
```diff
@@ -87,24 +87,37 @@ func (p *bmpeer) HandshakeComplete() bool {
...
+	p.peer.Disconnect()
 }
 
 // Start starts running the peer.
 func (p *bmpeer) Start() {
-	p.peer.Start()
+	peerLog.Infof(p.peer.PrependAddr("Started."))
+	if p.peer.Start() != nil {
+		p.server.donePeers <- p
+		peerLog.Infof(p.peer.PrependAddr("Failed to connect."))
```
Comment (ishbir) [inline peer.go]: Should be error.

File: peer.go [hunk 7]
```diff
@@ -87,24 +87,37 @@ func (p *bmpeer) HandshakeComplete() bool {
...
+func (p *bmpeer) Stop() {
+	peerLog.Debug(p.peer.PrependAddr("Stop called."))
+	// Only tell object manager we are gone if we ever told it we existed.
+	if p.HandshakeComplete() {
+		p.server.objectManager.DonePeer(p)
+	}
+
+	p.server.donePeers <- p
+	peerLog.Infof(p.peer.PrependAddr("Stopped."))
+}
```
Comment (ishbir) [inline peer.go]: Fix function doc. What's the point of logging Stop called and Stopped separately when there's only one pathway leading to the end? I'd say remove the debug log and just keep the info one.

File: peer.go [hunk 8]
```diff
@@ -325,8 +349,10 @@ func (p *bmpeer) HandleVersionMsg(msg *wire.MsgVersion) error {
 func (p *bmpeer) HandleVerAckMsg() error {
 	// If no version message has been sent disconnect.
 	if !p.versionSent {
+		peerLog.Infof(p.peer.PrependAddr("Ver ack msg received before version sent."))
```
Comment (ishbir) [inline peer.go]: Change to error.

File: peer.go [hunk 9]
```diff
@@ -447,6 +478,7 @@ func (p *bmpeer) handleInitialConnection() {
 		invVectList[i] = &wire.InvVect{Hash: hash}
 	}
 	p.PushInvMsg(invVectList)
+	peerLog.Infof(p.peer.PrependAddr("handshake complete."))
```
Comment (ishbir) [inline peer.go]: I think all the log entries related to messages and their contents should be of level trace since we're tracing the execution of bmd and its interaction with clients (almost all of log entries in this file). We don't want to clutter everything at level info.

File: peer/peer.go [hunk 10]
```diff
@@ -142,47 +143,47 @@ out:
 		// Stop the timer now, if we go around again we will reset it.
 		idleTimer.Stop()
 		if err != nil {
+			log.Errorf(p.PrependAddr("Disconnecting due to invalid message received: "), err)
+			break out
```
Comment (ishbir) [inline peer/peer.go]: I'm in a fix about this. Should we really disconnect clients for invalid messages sent? We put old clients at a risk for not being able to connect to the network. I don't think PyBitmessage does this. What do you think?

File: peer/peer.go [hunk 11]
```diff
@@ -142,47 +143,47 @@ out:
...
 		case *wire.MsgGetPubKey, *wire.MsgPubKey, *wire.MsgMsg, *wire.MsgBroadcast, *wire.MsgUnknownObject:
 			err = p.logic.HandleObjectMsg(rmsg)
-			markConnected = true
 
 		case *wire.MsgPong:
-			markConnected = true
+			
+		default:
+			log.Debug(p.PrependAddr("Invalid message processed. This should not happen."))
+			break out
```
Comment (ishbir) [inline peer/peer.go]: Then shouldn't the log level be critical or even better, a panic, since it isn't supposed to happen with even a malicious peer.

File: peer/send.go [hunk 12]
```diff
@@ -240,6 +241,8 @@ out:
 			if !send.Running() || invSendQueue.Len() == 0 {
 				continue
 			}
+			
+			log.Info(send.PrependAddr("Trickling an inv."))
```
Comment (ishbir) [inline peer/send.go]: Trace.

File: server.go [hunk 13]
```diff
@@ -36,7 +36,7 @@ const (
 	connectionRetryInterval = time.Second * 10
 
 	// defaultMaxOutbound is the default number of max outbound peers.
-	defaultMaxOutbound = 8
+	defaultMaxOutbound = 25
```
Comment (ishbir) [inline server.go]: 25 is a LOT. We would have to keep track of every peer's known inventory. This consumes a lot of memory. Why did you change it?

File: server.go [hunk 14]
```diff
@@ -286,6 +283,8 @@ type getAddedNodesMsg struct {
 // This function exists to add initial peers to the address manager before the
 // peerHandler go routine has entered its main loop.
 func (s *server) AddNewPeer(addr string, stream uint32, permanent bool) error {
+	serverLog.Info("Creating peer at ", addr, ", stream: ", stream)
+
```
Comment (ishbir) [inline server.go]: Better as debug.

File: server.go [hunk 15]
```diff
@@ -415,6 +414,7 @@ func (s *server) listenHandler(listener peer.Listener) {
 // peers to and from the server, banning peers, and broadcasting messages to
 // peers. It must be run in a goroutine.
 func (s *server) peerHandler() {
+	serverLog.Debug("Peer handler started.")
```
Comment (ishbir) [inline server.go]: Change to info.

File: peer.go (hunk 9 above)
Comment (DanielKrawisz) [inline peer.go]: Sorry, I've never used log levels before. I was just trying to find the bugs!

File: peer/peer.go (hunk 10 above)
Comment (DanielKrawisz) [inline peer/peer.go]: If we don't have a rule to deal with a particular kind of message, I don't see that there's anything else we can do. If we want to support older versions of the protocol, then we just have to tell bmd how to interpret the messages correctly.

File: server.go (hunk 13 above)
Comment (DanielKrawisz) [inline server.go]: I did this while I was playing around with the program and forgot to change it back. Do you think 8 is a good number?

File: peer/peer.go (hunk 10 above)
Comment (ishbir) [inline peer/peer.go]: But, for example, Atheros added a message with command "error", to aid in debugging. I don't think such custom messages should cause peers to be disconnected, but rather that message should just be ignored.

Comment (Kagami) [inline peer/peer.go]: Btw, I collected some of PyBitmessage networking specificities [here](https://github.com/bitchan/meta/wiki/PyBitmessage-networking-specificities) if you're interested.

Comment (DanielKrawisz) [inline peer/peer.go]: Oh thanks! Great, that's really helpful. And ok sure I'll set it to ignore message types we don't know about.

File: peer.go [hunk 16]
```diff
@@ -447,6 +474,7 @@ func (p *bmpeer) handleInitialConnection() {
 		invVectList[i] = &wire.InvVect{Hash: hash}
 	}
 	p.PushInvMsg(invVectList)
+	peerLog.Error(p.peer.PrependAddr("handshake complete."))
```
Comment (ishbir) [inline peer.go]: Error? Or debug?

File: config.go [hunk 17]
```diff
@@ -139,6 +141,7 @@ type config struct {
 	Upnp           bool          `long:"upnp" description:"Use UPnP to map our listening port outside of NAT"`
 	MaxUpPerPeer   Filesize      `long:"maxupload" description:"Maximum upload rate for any peer. Valid units are {B, K, M, G} bytes/sec."`
 	MaxDownPerPeer Filesize      `long:"maxdownload" description:"Maximum download rate for any peer. Valid units are {B, K, M, G} bytes/sec."`
+	MaxOutbound    int           `long:"maxoutbound" description:"The maximum number of outbound peers that bmd will try to maintain."`
```
Comment (ishbir) [inline config.go]: You didn't assign the default value of MaxOutbound.. should be done around line 335.

File: peer.go (hunk 16 above)
Comment (DanielKrawisz) [inline peer.go]: Yes, that should be debug, haha!
